Restructuring error reporting in start_traj_mode and start_point_queue_mode

jimmy-mcelwain opened the pull request:
Changes start_traj_mode and start_point_queue_mode services so that they properly report the reasons for failure in case motion cannot be started. This addresses many of the concerns brought up in the discussion of PR #265. I closed that PR and opened this to simplify bringing it up to date with main and to make the history less confusing.

jimmy-mcelwain commented:
Also in reference to #37 and #192

ted-miller commented on src/ControllerStatusIO.h:
```diff
@@ -128,7 +128,7 @@ extern BOOL Ros_Controller_IsPflActive();
 extern BOOL Ros_Controller_IsAnyFaultActive();
 extern BOOL Ros_Controller_MasterTaskIsJobName(const char* const jobName);
-extern MotionNotReadyCode Ros_Controller_GetNotReadySubcode();
+extern MotionNotReadyCode Ros_Controller_GetNotReadySubcode(bool attemptingToStartMotion);
```
The argument name doesn't match what's in the .c file

jimmy-mcelwain replied:
I'll fix that. The name in the .c file is what I intended to use

ted-miller commented on src/MotionControl.h:
```diff
@@ -12,6 +12,7 @@
 
 #define MOTION_START_TIMEOUT                5000  // in milliseconds
 #define MOTION_START_CHECK_PERIOD           50  // in millisecond
+#define MOTION_START_ERROR_MESSSAGE_LENGTH  256
```
```suggestion
#define MOTION_START_ERROR_MESSAGE_LENGTH  256
```

ted-miller commented on src/MotionControl.c:
```diff
@@ -1355,79 +1355,76 @@ static STATUS Ros_Controller_DisableEcoMode()
+                Ros_ErrorHandling_ErrNo_ToString(alarmcode),
+                alarmcode);
+            Ros_Debug_BroadcastMsg(output);
+            rosidl_runtime_c__String__assign(responseMessage, output);
```
Is it intentional to fall through and not return here?

jimmy-mcelwain replied:
Yes it is intentional. I can get rid of the fall through though and just duplicate the 2 lines if it's more clear that way.

ted-miller commented on src/MotionControl.c:
```diff
@@ -1355,79 +1355,76 @@ static STATUS Ros_Controller_DisableEcoMode()
+            return motion_readiness_code;
+        case MOTION_NOT_READY_ERROR:
+            alarmcode = Ros_Controller_GetAlarmCode();
+            snprintf(output, MOTION_START_ERROR_MESSSAGE_LENGTH, "%s: '%s' (0x%04X)",
```
I don't understand why this is the only case where the return string gets populated

ted-miller replied:
Ok. Just add some comment that states the message is being populated here because that this location has the most information about the error. Also, that other cases will populate the message upstream.

ted-miller commented on src/MotionControl.c:
```diff
@@ -1557,30 +1567,26 @@ BOOL Ros_MotionControl_StartMotionMode(MOTION_MODE mode)
 
     //Required to allow motion api to work (Potential race condition)
     Ros_Sleep(200);
+    motion_readiness_code = Ros_Controller_GetNotReadySubcode(false);
```
What's the purpose of the second call to `Ros_Controller_GetNotReadySubcode`?

jimmy-mcelwain replied:
The first call ignored "tractable problems" which is what I'm calling `MotionNotReady` codes that can be resolved within the `Ros_MotionControl_StartMotionMode()` function. This second call makes sure that it actually meets all of the conditions for motion. It is logically the same as a `Ros_Controller_IsMotionReady()` call, like the calls before it, but I just used `Ros_Controller_GetNotReadySubcode` instead just in case it failed so that the proper `motion_readiness_code` can be returned

ted-miller replied:
Ok. that makes sense. Can you add some comment to that effect?

gavanderhoorn commented:
@jimmy-mcelwain: could you perhaps describe your approach at a high-level?

You write:

> Changes start_traj_mode and start_point_queue_mode services so that they properly report the reasons for failure in case motion cannot be started.

what did they not do correctly before these changes, and which changes (to control flow perhaps) address that specifically?

jimmy-mcelwain commented:
What I wrote there doesn't describe what I changed since the last PR that I closed. I rebased on main and that comment sums up the purpose of this PR as a whole, not just the differences from the last PR.

This approach changes `Ros_Controller_GetNotReadySubcode()` to work better with `start_traj_mode` and `start_point_queue_mode` such that much less code is duplicated. Within `Ros_Controller_GetNotReadySubcode()` I check for intractable problems that require the user to take some action, e.g. like switching to remote mode or calling `/reset_error`. If I am attempting to start some motion mode, I will ignore tractable errors from `Ros_Controller_GetNotReadySubcode()` that are potentially addressed by starting the motion, e.g. turning on servos, starting the job. 

I also significantly simplified the control flow of the program. Previously, there were several calls to `goto updatestatus` which are no longer needed. The jumps to `[updatestatus](https://github.com/Yaskawa-Global/motoros2/blob/4133298e68a7dcf76ed6f636c79d922ddcdfda2d/src/MotionControl.c#L1560-L1583)` always followed some intractable issue. These `goto` calls didn't make sense to me, since they are checking if motion is ready almost immediately after they already determined that motion was not ready. I should have gotten rid of them in the previous PR. To me it makes a lot more sense to simply return with a `MotionNotReadyCode` if it is determined that the motion cannot start and nothing can be done about it.

ted-miller reviewed (approved):
testing worked well. Conflicts resolved using web editor
